cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -307,14 +321,17 @@ Result<std::unique_ptr<parquet::arrow::FileReader>> ParquetFileFormat::GetReader
   auto properties = MakeReaderProperties(*this, parquet_scan_options.get(), pool);
 
   ARROW_ASSIGN_OR_RAISE(auto input, source.Open());
-  std::unique_ptr<parquet::ParquetFileReader> reader;
-  try {
-    reader = parquet::ParquetFileReader::Open(std::move(input), std::move(properties));
-  } catch (const ::parquet::ParquetException& e) {
-    return Status::IOError("Could not open parquet input source '", source.path(),
-                           "': ", e.what());
-  }
 
+  auto maybe_reader = [&]() -> Result<std::unique_ptr<parquet::ParquetFileReader>> {
```

Comment:
Less fragile than macros would be:

```c++
  auto maybe_reader = parquet::CatchNotOk([&] {
    return parquet::ParquetFileReader::Open(std::move(input), std::move(properties));
  });
```

Could be implemented like:

```c++
template <typename Fn, typename Ret>
struct CatchNotOkImpl {
  static typename EnsureResult<Ret>::type Call(Fn&& fn) { return std::forward<Fn>(fn)(); }
};

template <typename Fn>
struct CatchNotOkImpl<Fn, Status> {
  static Status Call(Fn&& fn) { return std::forward<Fn>(fn)(); }
};

template <typename Fn>
struct CatchNotOkImpl<Fn, void> {
  static Status Call(Fn&& fn) {
    std::forward<Fn>(fn)();
    return Status::OK();
  }
};

template <typename Fn, typename Impl = CatchNotOkImpl<Fn, decltype(std::declval<Fn>()())>>
auto CatchNotOk(Fn&& fn) -> decltype(Impl::Call(std::forward<Fn>(fn))) try {
  return Impl::Call(std::forward<Fn>(fn));
} catch (const ParquetStatusException& e) {
  return e.status();
} catch (const ParquetException& e) {
  return Status::IOError(e.what());
}
```

What do you think?

Reply:
Hmm, I'm a bit lukewarm. Other similar functionality is already managed as macros (`ARROW_RETURN_NOT_OK`, `RETURN_IF_PYERROR`...). It seems like it would introduce a gratuitous variation. Also, given the size of the C++ metaprogramming alternative, this is possibly a reasonable use case for a macro.